File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.js
function solution(sequence, k) {
    let result = [0, 0];
    let minLength = Number.MAX_VALUE;
    let sum = 0;
    let start = 0;

    for (let end = 0; end < sequence.length; end++) {
        sum += sequence[end];

        while (sum > k) {
            sum -= sequence[start];
            start++;
        }

        if (sum === k && end - start < minLength) {
            minLength = end - start;
            result[0] = start;
            result[1] = end;
        }
    }

    return minLength === Number.MAX_VALUE ? [] : result;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java
import java.util.*;
class Solution {
    public int[] solution(int[] sequence, int k) {
        int[] answer = new int[2];
        int sum = 0;
        int start = 0;
        int length = Integer.MAX_VALUE;
        for (int end = 0; end < sequence.length; end++) {
            sum += sequence[end];
            // 함이 k랑 같다면
            if (sum == k) {
                // 길이가 작으면
                if (length > (end - start)) {
                    answer[0] = start;
                    answer[1] = end;
                    length = (end - start);
                }
            }
            else if (sum > k) {
                for (int j = start; j <= end; j++) {
                    sum -= sequence[j];
                    if (sum <= k) {
                        start = j + 1;
                        break;
                    }
                }
                if (sum == k) {
                    if (length > (end - start)) {
                        answer[0] = start;
                        answer[1] = end;
                        length = end - start;
                    }
                }
            }
        }
        System.out.println(Arrays.toString(answer));
        return answer;
    }
}